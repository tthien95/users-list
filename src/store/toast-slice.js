import { createSlice } from '@reduxjs/toolkit';

const toastSlice = createSlice({
  name: 'toast',
  initialState: { notification: null },
  reducers: {
    showNotification: (state, action) => {
      const { status, title, message } = action.payload;
      state.notification = { status, title, message };
    },
    hideNotification: (state) => {
      state.notification = null;
    }
  }
});

export const toastActions = toastSlice.actions;

export default toastSlice.reducer;
