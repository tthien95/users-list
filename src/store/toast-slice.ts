import { createSlice } from '@reduxjs/toolkit';

export interface ToastState {
  notification: {
    status: 'success' | 'error';
    title: string;
    message: string;
  } | null;
}

const initialState: ToastState = { notification: null };

const toastSlice = createSlice({
  name: 'toast',
  initialState,
  reducers: {
    showNotification: (state, action) => {
      const { status, title, message } = action.payload;
      state.notification = { status, title, message };
    },
    hideNotification: (state) => {
      state.notification = null;
    }
  }
});

export const toastActions = toastSlice.actions;

export default toastSlice.reducer;
