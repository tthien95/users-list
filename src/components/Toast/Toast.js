import React, { useEffect } from 'react';
import { createPortal } from 'react-dom';
import { useSelector, useDispatch } from 'react-redux';
import classes from './Toast.module.css';
import { toastActions } from '../../store/toast-slice';

const notiStatus = {
  success: '#5cb85c',
  error: '#d9534f'
};

const Toast = () => {
  const notification = useSelector((state) => state.notification);
  const dispatch = useDispatch();

  useEffect(() => {
    if (notification) {
      const interval = setTimeout(() => {
        dispatch(toastActions.hideNotification());
      }, 3000);

      return () => {
        clearTimeout(interval);
      };
    }
  }, [notification, dispatch]);

  if (!notification) {
    return null;
  }

  const { status, title, message } = notification;

  const content = (
    <div className={classes['notification-container']}>
      <div
        className={`${classes.notification} ${classes.toast}`}
        style={{ backgroundColor: notiStatus[status] }}
      >
        <button
          onClick={() => {
            dispatch(toastActions.hideNotification());
          }}
        >
          X
        </button>
        <div>
          <p className={classes['notification-title']}>{title}</p>
          <p className={classes['notification-message']}>{message}</p>
        </div>
      </div>
    </div>
  );

  return createPortal(content, document.getElementById('toastNoti'));
};

export default Toast;
