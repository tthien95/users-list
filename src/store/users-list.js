import React, { useState, useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { get } from '../utils/api-helper';
import { toastActions } from './toast-slice';

const UsersListContext = React.createContext({
  usersList: [],
  isLoading: false,
  addUser: (value) => {},
  updateUser: (value) => {}
});

export const UserContextProvider = (props) => {
  const [usersList, setUsersList] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const dispatch = useDispatch();

  useEffect(() => {
    setIsLoading(true);

    const fnHandleError = ({ response }) => {
      dispatch(
        toastActions.showNotification({
          status: 'error',
          title: 'Error',
          message:
            response?.data?.message ||
            response?.statusText ||
            'There is something wrong happended while fetching data'
        })
      );
    };

    get('/users')
      .then((res) => {
        const usersList = res.data.users;
        setUsersList(usersList);
      }, fnHandleError)
      .finally(() => {
        setIsLoading(false);
      });
  }, [dispatch]);

  const addUser = (value) => {
    let currUsersList = [...usersList];
    currUsersList.unshift(value);
    setUsersList(currUsersList);
  };

  const updateUser = (value) => {
    let currUsersList = [...usersList];
    // eslint-disable-next-line eqeqeq
    const userIndx = currUsersList.findIndex(({ id }) => id == value.id);
    if (userIndx > -1) {
      currUsersList.splice(userIndx, 1, { ...value });
      setUsersList(currUsersList);
    }
  };

  return (
    <UsersListContext.Provider
      value={{ usersList, addUser, updateUser, isLoading }}
    >
      {props.children}
    </UsersListContext.Provider>
  );
};

export default UsersListContext;
