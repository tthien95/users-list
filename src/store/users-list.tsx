import React, { useState, useCallback } from 'react';
import { useDispatch } from 'react-redux';
import { toastActions } from './toast-slice';
import { User, UsersListContextType } from '../type/user';
import { AxiosError } from 'axios';

const UsersListContext = React.createContext<UsersListContextType>({
  usersList: [],
  setUsersList: () => {},
  isLoading: false,
  addUser: () => {},
  updateUser: () => {},
  deleteUser: () => {},
  setIsLoading: () => {},
  fnHandleError: () => {}
});

type ErrorDataRes = {
  message: string;
};

const doesDataContainMessage = (data: unknown): data is ErrorDataRes => {
  return typeof data === 'object' && data !== null && 'message' in data;
};

export const UserContextProvider: React.FC<React.PropsWithChildren> = ({
  children
}) => {
  const [usersList, setUsersList] = useState<User[] | []>([]);
  const [isLoading, setIsLoading] = useState(false);
  const dispatch = useDispatch();

  const fnHandleError = useCallback(
    ({ response }: AxiosError) => {
      dispatch(
        toastActions.showNotification({
          status: 'error',
          title: 'Error',
          message:
            (doesDataContainMessage(response?.data) &&
              response?.data.message) ||
            response?.statusText ||
            'There is something wrong happended while fetching data'
        })
      );
    },
    [dispatch]
  );

  const addUser = (value: User) => {
    let currUsersList = [...usersList];
    currUsersList.unshift(value);
    setUsersList(currUsersList);
  };

  const updateUser = (value: User) => {
    let currUsersList = [...usersList];
    // eslint-disable-next-line eqeqeq
    const userIndx = currUsersList.findIndex(({ id }) => id == value.id);
    if (userIndx > -1) {
      currUsersList.splice(userIndx, 1, value);
      setUsersList(currUsersList);
    }
  };

  const deleteUser = (userId: string) => {
    let currUsersList = [...usersList];
    // eslint-disable-next-line eqeqeq
    const userIndx = currUsersList.findIndex(({ id }) => id == userId);
    if (userIndx > -1) {
      currUsersList.splice(userIndx, 1);
      setUsersList(currUsersList);
    }
  };

  return (
    <UsersListContext.Provider
      value={{
        usersList,
        setUsersList,
        addUser,
        updateUser,
        deleteUser,
        isLoading,
        setIsLoading,
        fnHandleError
      }}
    >
      {children}
    </UsersListContext.Provider>
  );
};

export default UsersListContext;
