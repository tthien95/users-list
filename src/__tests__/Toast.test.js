import { cleanup, render, screen } from '@testing-library/react';
import Toast from '../components/Toast/Toast';

import * as Redux from 'react-redux';
import store from '../store/index';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn()
}));

jest.mock('react-dom', () => ({
  ...jest.requireActual('react-dom'),
  createPortal: jest.fn((component, container) => component)
}));

const renderWithProvider = () => {
  return render(
    <Redux.Provider store={store}>
      <Toast />
    </Redux.Provider>
  );
};

describe('Navigation', () => {
  afterAll(cleanup);

  it('should match snapshot for success', () => {
    Redux.useSelector.mockReturnValueOnce({
      status: 'success',
      title: 'Success',
      message: 'Success Message'
    });
    const { asFragment } = renderWithProvider();
    expect(asFragment(<Toast />)).toMatchSnapshot();
  });

  it('should match snapshot for error', () => {
    Redux.useSelector.mockReturnValueOnce({
      status: 'error',
      title: 'error',
      message: 'Error Message'
    });
    const { asFragment } = renderWithProvider();
    expect(asFragment(<Toast />)).toMatchSnapshot();
  });

  it('should not render toast when there is no notification', () => {
    Redux.useSelector.mockReturnValueOnce(null);
    expect(screen.queryByRole('alert')).not.toBeInTheDocument();
  });
});
