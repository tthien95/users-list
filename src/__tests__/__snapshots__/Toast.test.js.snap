// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation should match snapshot for error 1`] = `<DocumentFragment />`;

exports[`Navigation should match snapshot for success 1`] = `<DocumentFragment />`;
